### PR TITLE
Treasure Chest Game Shuffle

### PR DESCRIPTION
**Pull request by Jaybone25**
Gives the ability to shuffle treasure chest game keys and chests.   Has all the options of other keysanity modes.

### CODE REVIEW

File: ASM/c/chests.c [hunk 1]
```diff
@@ -39,15 +40,10 @@ void get_chest_override(z64_actor_t *actor) {
             }
             if (CHEST_SIZE_MATCH_CONTENTS || CHEST_SIZE_TEXTURE) {
                 if (item_row->chest_type == BROWN_CHEST || item_row->chest_type == SILVER_CHEST || item_row->chest_type == SKULL_CHEST_SMALL || item_row->chest_type == HEART_CHEST_SMALL) {
-                    // Small chest
-                    size = SMALL_CHEST;
-                }
-                else {
```
Comment (fenhl) [inline ASM/c/chests.c:45]: This `else` block needs to be restored.

File: ASM/c/dungeon_info.c [hunk 2]
```diff
@@ -208,11 +211,12 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
         }
         text_flush_size(db, font_width, font_height, 0, 0);
 
-        left += (8 * font_width) + padding;
+        left += (11 * font_width) + padding;
 
         // Draw keys
 
         if (show_keys) {
+
```
Comment (fenhl) [inline ASM/c/dungeon_info.c]: Please don't introduce arbitrary whitespace changes.

File: ASM/c/dungeon_info.c [hunk 3]
```diff
@@ -230,8 +234,10 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
                 if (total_keys > 9) total_keys = 9;
 
                 char count[5] = "O(O)";
-                if (current_keys > 0) count[0] = current_keys + '0';
-                if (total_keys > 0) count[2] = total_keys + '0';
+                if (SHUFFLE_CHEST_GAME) {
```
Comment (fenhl) [inline ASM/c/dungeon_info.c]: It would probably be better to only show the chest game row if this setting is on. In any case, this conditional doesn't make sense as it's being applied to all rows.

File: ASM/c/dungeon_info.c [hunk 4]
```diff
@@ -478,8 +484,10 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
...
```
Comment (fenhl) [inline ASM/c/dungeon_info.c]: Same thing here.

File: ItemPool.py [hunk 5]
```diff
@@ -612,6 +619,20 @@ def get_pool_core(world):
             if shuffle_item and world.settings.gerudo_fortress == 'normal' and 'Thieves Hideout' in world.settings.key_rings:
                 item = get_junk_item()[0] if location.name != 'Hideout 1 Torch Jail Gerudo Key' else 'Small Key Ring (Thieves Hideout)'
 
+        # Treasure Chest Game Key Shuffle
+        elif location.vanilla_item in ['Small Key (Treasure Chest Game)', 'Green Rupee (Treasure Chest Game)', 'Blue Rupee (Treasure Chest Game)', 'Red Rupee (Treasure Chest Game)'] and location.scene == 0x10:
```
Comment (fenhl) [inline ItemPool.py]:
```suggestion
        elif location.scene == 0x10 and location.vanilla_item != 'Piece of Heart (Treasure Chest Game)':
```

File: Patches.py [hunk 6]
```diff
@@ -2112,6 +2120,10 @@ def update_scrub_text(message, text_replacement, default_price, price, item_name
     SKULL_CHEST_BIG =  15
     HEART_CHEST_SMALL = 16
     HEART_CHEST_BIG = 17
+    if world.settings.shuffle_tcgkeys == 'vanilla':
+        item = read_rom_item(rom, 0x71)
```
Comment (fenhl) [inline Patches.py:2155]: This could use a comment explaining what it's doing (changing chest game keys to be in brown chests).

File: Patches.py [hunk 7]
```diff
@@ -2913,3 +2944,5 @@ def patch_ganons_tower_bk_door(rom: Rom, flag):
     var = (0x05 << 6) + (flag & 0x3F)
     bytes = [(var & 0xFF00) >> 8, var & 0xFF]
     rom.write_bytes(0x2EE30FE, bytes)
+
```
Comment (fenhl) [inline Patches.py]: More arbitrary whitespace changes.

File: SaveContext.py [hunk 8]
```diff
@@ -1045,7 +1053,10 @@ def get_save_context_addresses(self):
             'double_magic'          : [False, True],
         },
         "Rupee"                     : {'rupees' : None},
-        "Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Green Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Blue Rupee (Treasure Chest Game)" : {'rupees' : None},
```
Comment (fenhl) [inline SaveContext.py]: These need to be changed from `None` to their actual values, since the values can't be derived from the item names.

File: SettingsList.py [hunk 9]
```diff
@@ -3005,15 +3005,57 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
     Combobox(
         name           = 'key_rings_choice',
         gui_text       = 'Key Rings Mode',
         default        = 'off',
         choices        = {
             'off':       'Off',
-            'choice':    'Choose dungeons',
-            'all':       'All dungeons',
-            'random':    'Random dungeons'
+            'choice':    'Choose dungeons/Treasure Chest Game',
```
Comment (fenhl) [inline SettingsList.py]: If you're going to change this, also add Thieves' Hideout here. I'd say it's not necessary in the option names and should probably be mentioned in the tooltip instead.

File: data/World/Overworld.json [hunk 10]
```diff
@@ -1385,8 +1385,31 @@
     {
         "region_name": "Market Treasure Chest Game",
         "scene": "Market Treasure Chest Game",
+        "hint": "MARKET",
         "locations": {
-            "Market Treasure Chest Game Reward": "can_use(Lens_of_Truth)"
+            "Market Treasure Chest Game Salesman": "is_child and at_night",
+            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys == 'vanilla' and can_use(Lens_of_Truth)) or (shuffle_tcgkeys != 'vanilla' and
```
Comment (fenhl) [inline data/World/Overworld.json]: These can be massively simplified.

* You don't need to check for keysy, that's handled by logically giving the keys as starting items.
* You don't need to check for vanilla other than to see if lens is required. Let the logic assume that you can open both chests, which ends up being equivalent.

For example, I'd write the logic for the reward as:

```python
(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth)) and (Small_Key_Treasure_Chest_Game, 6)
```

**Review by fenhl — COMMENTED** [review]
In `Notes/extra-save-data.md`, also update the note about the total small key count in the unused field to say `Scenes 0x00–0x10`.

File: data/World/Overworld.json [hunk 11]
```diff
@@ -1385,8 +1385,31 @@
     {
         "region_name": "Market Treasure Chest Game",
         "scene": "Market Treasure Chest Game",
+        "hint": "MARKET",
```
Comment (fenhl) [inline data/World/Overworld.json]: This changes the treasure chest game to be considered part of the market hint area regardless of where it's shuffled to. I think this is unintuitive due to its inconsistency with other interiors and should be reverted.

Comment (Jaybone25) [reply data/World/Overworld.json]: This was to troubleshoot a seed generation error due to hint region.   We fixed it and forgot to remove it, we'll remove it

File: ItemPool.py (hunk 5 above)
Comment (Jaybone25) [reply ItemPool.py]: Fixed

File: Patches.py (hunk 7 above)
Comment (Jaybone25) [reply Patches.py]: There were actually 3 lines of white space here.   Removed them all.

File: SaveContext.py (hunk 8 above)
Comment (Jaybone25) [reply SaveContext.py]: Done

File: ItemPool.py (hunk 5 above)
Comment (fenhl) [reply ItemPool.py]: Did you forget to push the changes?

File: ASM/c/dungeon_info.c (hunk 2 above)
Comment (Jaybone25) [reply ASM/c/dungeon_info.c]: Removed whitespace, however all existing ones had the same white space.   I did not adjust those

File: Patches.py [hunk 12]
```diff
@@ -2943,6 +2943,4 @@ def patch_rupee_tower(location, rom: Rom):
 def patch_ganons_tower_bk_door(rom: Rom, flag):
     var = (0x05 << 6) + (flag & 0x3F)
     bytes = [(var & 0xFF00) >> 8, var & 0xFF]
-    rom.write_bytes(0x2EE30FE, bytes)
-
-
+    rom.write_bytes(0x2EE30FE, bytes)
```
Comment (fenhl) [inline Patches.py]: You deleted one newline too many. By convention, all text files end with exactly one newline.

File: data/World/Overworld.json [hunk 13]
```diff
@@ -1387,7 +1387,7 @@
         "scene": "Market Treasure Chest Game",
         "hint": "MARKET",
         "locations": {
-            "Market Treasure Chest Game Salesman": "is_child and at_night",
+            "Market Treasure Chest Game Salesman": true,
```
Comment (fenhl) [inline data/World/Overworld.json]:
```suggestion
            "Market Treasure Chest Game Salesman": "True",
```

File: ASM/c/dungeon_info.c (hunk 3 above)
Comment (jdunn596) [reply ASM/c/dungeon_info.c]: That is actually the way I had it before but because of the way the symbol is written to the rom, it will still show when keysy is on. But I see what you mean otherwise and ill fix it

File: ASM/c/dungeon_info.c (hunk 4 above)
Comment (jdunn596) [reply ASM/c/dungeon_info.c]: Same as my reply above with this

File: ASM/c/dungeon_info.c (hunk 3 above)
Comment (fenhl) [reply ASM/c/dungeon_info.c]: You could set the symbol to `2` in keysy and compare it to `1` here.

Comment (jdunn596) [reply ASM/c/dungeon_info.c]: oh so in patches.py, Could I refactor the code to say:

```
if world.settings.shuffle_tcgkeys != 'vanilla':
        if 'keysy' in world.settings.shuffle_tcgkeys:
                rom.write_byte(rom.sym('SHUFFLE_CHEST_GAME'), 0x02)
        else:
                rom.write_byte(rom.sym('SHUFFLE_CHEST_GAME'), 0x01)
        # Update Chest Game Salesman to better fit the fact he sells a randomized item
        update_message_by_id(messages, 0x6D, "I seem to have misplaced my\x01keys, but I have a fun item to\x01sell instead.\x04How about \x05\x4110 Rupees\x05\x40?\x01\x01\x1B\x05\x42Buy\x01Don't Buy\x05\x40\x02")
        update_message_by_id(messages, 0x2D, "That's OK!\x01More fun for me.\x02")
        update_message_by_id(messages, 0x6E, "Wait, that room was off limits!\x02")
        update_message_by_id(messages, 0x704C, "I hope you like it!\x02")
```

Sorry for the questions, I'm still learning lol

Comment (fenhl) [reply ASM/c/dungeon_info.c]: `world.settings.shuffle_tcgkeys == 'keysy'`

File: ASM/c/chests.c (hunk 1 above)
Comment (jdunn596) [reply ASM/c/chests.c:45]: addressed

File: ASM/c/dungeon_info.c (hunk 3 above)
Comment (jdunn596) [reply ASM/c/dungeon_info.c]: addressed

File: ASM/c/dungeon_info.c (hunk 4 above)
Comment (jdunn596) [reply ASM/c/dungeon_info.c]: addressed

File: ItemPool.py (hunk 5 above)
Comment (jdunn596) [reply ItemPool.py]: addressed

File: Patches.py (hunk 6 above)
Comment (jdunn596) [reply Patches.py:2155]: addressed

File: SettingsList.py (hunk 9 above)
Comment (jdunn596) [reply SettingsList.py]: addressed

File: data/World/Overworld.json (hunk 10 above)
Comment (jdunn596) [reply data/World/Overworld.json]: addressed

File: data/World/Overworld.json (hunk 11 above)
Comment (jdunn596) [reply data/World/Overworld.json]: addressed

File: Patches.py (hunk 12 above)
Comment (jdunn596) [reply Patches.py]: addressed

**Comment by jdunn596** [comment]
All issues that were pointed out _should_ be resolved. I'll wait to hear back :D

File: ASM/c/dungeon_info.c [hunk 14]
```diff
@@ -110,7 +113,7 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
         int font_width = 6;
         int font_height = 11;
         int padding = 1;
-        int rows = 13;
+        int rows = (!SHUFFLE_CHEST_GAME || SHUFFLE_CHEST_GAME == 2) ? 13 : 14;
```
Comment (fenhl) [inline ASM/c/dungeon_info.c]: Can be simplified to:

```suggestion
        int rows = SHUFFLE_CHEST_GAME == 1 ? 14 : 13;
```

File: ASM/c/dungeon_info.c [hunk 15]
```diff
@@ -201,39 +204,70 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
 
         // Draw dungeon names
 
-        for (int i = 0; i < dungeon_count; i++) {
-            dungeon_entry_t *d = &(dungeons[i]);
-            int top = start_top + ((icon_size + padding) * i) + 1;
-            text_print_size(d->name, left, top, font_width);
-        }
-        text_flush_size(db, font_width, font_height, 0, 0);
+        if (!SHUFFLE_CHEST_GAME || SHUFFLE_CHEST_GAME == 2) {
```
Comment (fenhl) [inline ASM/c/dungeon_info.c]: You could merge the “then” and “else” blocks since they're so similar, only checking for `SHUFFLE_CHEST_GAME == 1` in the two places where they should differ, e.g. `i < dungeon_count - (SHUFFLE_CHEST_GAME == 1 ? 0 : 1)`. Same thing for more `if` statements below.

File: SaveContext.py [hunk 16]
```diff
@@ -1045,7 +1053,10 @@ def get_save_context_addresses(self):
             'double_magic'          : [False, True],
         },
         "Rupee"                     : {'rupees' : None},
-        "Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Green Rupee (Treasure Chest Game)" : {'rupees' : 1},
+        "Blue Rupee (Treasure Chest Game)" : {'rupees' : 5},
+        "Red Rupee (Treasure Chest Game)" : {'rupees' : 10},
```
Comment (fenhl) [inline SaveContext.py]:
```suggestion
        "Red Rupee (Treasure Chest Game)" : {'rupees' : 20},
```

File: data/World/Overworld.json [hunk 17]
```diff
@@ -1386,7 +1386,18 @@
         "region_name": "Market Treasure Chest Game",
         "scene": "Market Treasure Chest Game",
         "locations": {
-            "Market Treasure Chest Game Reward": "can_use(Lens_of_Truth)"
+            "Market Treasure Chest Game Salesman": "True",
+            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth) and (Small_Key_Treasure_Chest_Game, 6))",
```
Comment (fenhl) [inline data/World/Overworld.json]:
```suggestion
            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth)) and (Small_Key_Treasure_Chest_Game, 6)",
```

File: data/World/Overworld.json [hunk 18]
```diff
@@ -1386,7 +1386,18 @@
...
+            "Market Treasure Chest Game Room 1 Bottom": "(shuffle_tcgkeys != 'vanilla' and (Small_Key_Treasure_Chest_Game, 1))",
```
Comment (fenhl) [inline data/World/Overworld.json]:
```suggestion
            "Market Treasure Chest Game Room 1 Bottom": "(Small_Key_Treasure_Chest_Game, 1)",
```

Similar for the other 9 locations.

File: ItemPool.py [hunk 19]
```diff
@@ -612,6 +619,20 @@ def get_pool_core(world):
...
                 item = get_junk_item()[0] if location.name != 'Hideout 1 Torch Jail Gerudo Key' else 'Small Key Ring (Thieves Hideout)'
 
+        # Treasure Chest Game Key Shuffle
+        elif location.vanilla_item != 'Piece of Heart (Treasure Chest Game)' and location.scene == 0x10:
+            if world.settings.shuffle_tcgkeys in ['regional', 'overworld', 'any_dungeon', 'keysanity']:
+                if 'Treasure Chest Game' in world.settings.key_rings and location.vanilla_item == 'Small Key (Treasure Chest Game)':
+                    item = get_junk_item()[0] if location.name != 'Market Treasure Chest Game Salesman' else 'Small Key Ring (Treasure Chest Game)'
+                shuffle_item = True
+            elif world.settings.shuffle_tcgkeys == 'remove' and location.vanilla_item == 'Small Key (Treasure Chest Game)':
+                item = get_junk_item()[0]
```
Comment (fenhl) [inline ItemPool.py]: Keysy is implemented by logically starting with all the keys:

```python
world.state.collect(ItemFactory(item))
```

Add that line above this line so `item` isn't changed to a junk item yet.

**Comment by Jaybone25** [comment]
All suggested changes have been made.

**Comment by fenhl** [comment]
I'll give this another round of review soonish.

**Comment by r0bd0g** [comment]
For some reason, only Room 1 bottom doesn't check for shuffle_tcgkeys != 'vanilla', which seems strange since all the others do. I think it was a suggested change that said to apply it to the other 9 locations, but this wasn't done.

It's a bit hard to follow so I just want to make sure, how is the logic getting the 6 keys to reach the final reward when chest game is vanilla? If all of the chests are checking for shuffling to not be vanilla, if you're relying on getting the keys from vanilla locations, that won't work. I checked also, maybe since the keys are irrelevant, you decided to logically start with them, but I don't see that anywhere either. So I'm confused how to logic is to reach the final reward in vanilla behaviour.

I worry a bit if the intended plan is to collect the keys one at a time from their vanilla locations, if the spoiler log really needs to have like 6 spheres dedicated to collecting and using the vanilla keys. I know that vanilla keys in other places do do that, but the vanilla behaviour of these vanilla key locations is a lot different from the usual vanilla keys, so I'm not sure displaying them in the spoiler log one at a time like that is appropriate.

I really think the issues with ice traps in the treasure chest game ought to be solved.

**Comment by jdunn596** [comment]
> For some reason, only Room 1 bottom doesn't check for shuffle_tcgkeys != 'vanilla', which seems strange since all the others do. I think it was a suggested change that said to apply it to the other 9 locations, but this wasn't done.
> 
> It's a bit hard to follow so I just want to make sure, how is the logic getting the 6 keys to reach the final reward when chest game is vanilla? If all of the chests are checking for shuffling to not be vanilla, if you're relying on getting the keys from vanilla locations, that won't work. I checked also, maybe since the keys are irrelevant, you decided to logically start with them, but I don't see that anywhere either. So I'm confused how to logic is to reach the final reward in vanilla behaviour.
> 
> I worry a bit if the intended plan is to collect the keys one at a time from their vanilla locations, if the spoiler log really needs to have like 6 spheres dedicated to collecting and using the vanilla keys. I know that vanilla keys in other places do do that, but the vanilla behaviour of these vanilla key locations is a lot different from the usual vanilla keys, so I'm not sure displaying them in the spoiler log one at a time like that is appropriate.

Good catch on the logic file. I have no idea what happened with that but I assume something happened between both Jay and I merging changes but that should now be fixed. EDIT: Okay now it's _actually_ fixed

> I really think the issues with ice traps in the treasure chest game ought to be solved.

In regard to ice traps, we decided to just extend fenhl's method to store them in a queue until you leave the scene before they are applied. You can open them from a chest in the minigame. The textbox will display 'You are a FOOL', and you can continue to 
open the rest of the chests available to you. Upon leaving/save warping, all traps in the queue will be applied. There's really not much we can do about this until we better understand why the crashes take place. If you have any ideas, I'm open to suggestions.

**Comment by r0bd0g** [comment]
I'm not sure whether the intention was to add the settings check back to the one chest that was missing it, or to remove it from the other 9 that still had it. Maybe Fenhl can explain what they intended there?

That doesn't work for the final chest. It lets you through with lens even when it's not vanilla behaviour. I think you need not vanilla and keys, or vanilla and lens. (Then the settings checks on all the individual chests won't be necessary since who cares if the logic picks up some vanilla keys or loser rupees? Similar to how the first key that you buy is not locked by any settings check either. But again I want to know what Fenhl intended with the suggestion.)

**Comment by jdunn596** [comment]
> I'm not sure whether the intention was to add the settings check back to the one chest that was missing it, or to remove it from the other 9 that still had it. Maybe Fenhl can explain what he intends there?
> 
> That doesn't work for the final chest. It lets you through with lens even when it's not vanilla behaviour. I think you need not vanilla and keys, or vanilla and lens. (Then the settings checks on all the individual chests won't be necessary since who cares if the logic picks up some vanilla keys or loser rupees? But again I want to know what Fenhl intended with the suggestion.)

Yeah that's actually what we had originally but I guess we'll just wait for it to be reviewed again and see what fenhl thinks. Nbd if it needs to be changed in the long run

**Comment by jdunn596** [comment]
You were right r0b I tested with a spoiler log and it was indeed accounting for lens for all settings. The adjustment should handle it

File: LocationList.py [hunk 20]
```diff
@@ -1855,12 +1866,12 @@ def shop_address(shop_id, shelf_id):
     ("Gerudo Training Ground MQ Second Iron Knuckle Chest",          ("Chest",        0x0B,  0x12, None,                            'Arrows (10)',                           ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
     ("Gerudo Training Ground MQ Flame Circle Chest",                 ("Chest",        0x0B,  0x0E, None,                            'Small Key (Gerudo Training Ground)',    ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
     ("Gerudo Training Ground MQ Maze Right Central Chest",           ("Chest",        0x0B,  0x05, None,                            'Rupees (5)',                            ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
-    ("Gerudo Training Ground MQ Maze Right Side Chest",              ("Chest",        0x0B,  0x08, None,                            'Rupee (Treasure Chest Game)',           ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
+    ("Gerudo Training Ground MQ Maze Right Side Chest",              ("Chest",        0x0B,  0x08, None,                            'Green Rupee (Treasure Chest Game)',     ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
```
Comment (mzxrules) [inline LocationList.py]: nitpick, but should this be Rupee (1) (Treasure Chest Game) to match other instances?

File: Patches.py [hunk 21]
```diff
@@ -2384,6 +2400,25 @@ def update_scrub_text(message, text_replacement, default_price, price, item_name
     save_context.equip_current_items(world.settings.starting_age)
     save_context.write_save_table(rom)
 
+    # Convert temporary flags used for locked doors in Treasure Chest Game to permanent flags namely (0x1A-0x1F)
+    if world.settings.shuffle_tcgkeys != 'vanilla':
+        rom.write_byte(0x33A6070 +15, 0xDF)
```
Comment (mzxrules) [inline Patches.py]:
```suggestion
        rom.write_byte(0x33A607F, 0xDF)
```

repeat for all

File: LocationList.py (hunk 20 above)
Comment (fenhl) [reply LocationList.py]: I think `Rupee (Treasure Chest Game) (1)` would be best as it would work automatically with the starting items code that translates item names in that format into ammo amounts.

**Comment by jdunn596** [comment]
suggested changes have been made in the most recent commit

**Comment by fenhl** [comment]
> Maybe Fenhl can explain what he intends there?

They/them pronouns for me please. As for the question itself, I'm going to send a PR once this PR is rebased onto latest Dev, since that's easier than trying to explain it in a review.

**Comment by fenhl** [comment]
Jaybone25#6

**Comment by r0bd0g** [comment]
Logic-wise the only thing I can really say about this implementation is that technically the chest that isn't assigned as the vanilla location doesn't need all that extra logic.

**Comment by fenhl** [comment]
I think the logic is clearer this way, since the chests are arbitrarily assigned rupees and keys as a hack.

**Comment by r0bd0g** [comment]
I think you should comment in the logic file which chests actually have the vanilla keys.

**Comment by Jaybone25** [comment]
I don’t think that makes sense since in vanilla they are random.For the purposes of the hack I just removed the ability for it to place one on the top chest, but vanilla still behaves like normal.Sent from my iPhoneOn Apr 25, 2023, at 8:20 AM, r0bd0g ***@***.***> wrote:﻿
I think you should comment in the logic file which chests actually have the vanilla keys.

**Comment by r0bd0g** [comment]
I mean I think the logic files should be aware of the difference between top and bottom somehow. Otherwise it's kind of stealth atm.

**Comment by cjohnson57** [comment]
Lmao
image

**Comment by Jaybone25** [comment]
Unfortunately that line will rarely be seen lol

**Review by cjohnson57 — APPROVED** [review]
One non-blocking issue, if you can fix before I merge that's great but it won't stop me from merging:
If you say "No" to buying his item, you need to reload the scene to be able to talk to him again.

**Comment by Jaybone25** [comment]
> 

Will try, I attempted it before w/o success, but I have an hour to take a look again
